cassandra/src/test/java/brave/cassandra/ITTracing.java

```diff
@@ -116,13 +137,57 @@ void execute(Function<Session, BoundStatement> statement) {
-        .build(); Session session = TracingSession.create(withPropagation, cluster.connect())) {
-      session.execute(statement.apply(session));
+  static ImageFromDockerfile imageWithBraveTracer() {
+    ImageFromDockerfile image = new ImageFromDockerfile("openzipkin/brave-cassandra:test");
+    // First detect if we are in an IDE or failsafe. The latter will see our shaded jar.
+    MountableFile tracingCodeSource = codeSource(Tracing.class);
+    if (tracingCodeSource.getResolvedPath().contains("brave-instrumentation-cassandra")) {
+      return image
+          .withFileFromTransferable("brave-instrumentation-cassandra.jar", tracingCodeSource)
+          .withDockerfileFromBuilder(
+              builder -> builder.from(CassandraContainer.IMAGE_NAME.asCanonicalNameString())
+                  .add("brave-instrumentation-cassandra.jar",
+                      "lib/brave-instrumentation-cassandra.jar")
+          );
     }
+
+    // Otherwise, we need references to our main classpath in Cassandra's classpath
+    return image
+        .withFileFromTransferable("classes", tracingCodeSource)
+        .withFileFromTransferable("brave.jar", codeSource(brave.Tracing.class))
+        .withFileFromTransferable("zipkin-reporter-brave.jar", codeSource(ZipkinSpanHandler.class))
+        .withFileFromTransferable("zipkin-sender-urlconnection.jar",
+            codeSource(URLConnectionSender.class))
+        .withFileFromTransferable("zipkin-reporter.jar", codeSource(Reporter.class))
+        .withFileFromTransferable("zipkin.jar", codeSource(zipkin2.Span.class))
+        .withDockerfileFromBuilder(
+            builder -> builder.from(CassandraContainer.IMAGE_NAME.asCanonicalNameString())
+                // Copy the above references to a new layer over Zipkin's Cassandra image
+                .add("classes/", "classes/")
+                .add("brave.jar", "lib/brave.jar")
+                .add("zipkin-reporter-brave.jar", "lib/zipkin-reporter-brave.jar")
+                .add("zipkin-sender-urlconnection.jar", "lib/zipkin-sender-urlconnection.jar")
+                .add("zipkin-reporter.jar", "lib/zipkin-reporter.jar")
+                .add("zipkin.jar", "lib/zipkin.jar")
+        );
+  }
+
+  static MountableFile codeSource(Class<?> clazz) {
+    return MountableFile.forHostPath(
+        clazz.getProtectionDomain().getCodeSource().getLocation().getFile());
+  }
+
+  /** Overwrite JAVA_OPTS to enable tracing and point it at the test Zipkin endpoint */
+  @NotNull private static String javaOpts(int zipkinHttpPort) {
+    // TODO: would be nicer if Testcontainers.exposeHostPort(int) and returned the input
+    // https://github.com/testcontainers/testcontainers-java/issues/3538
+    Testcontainers.exposeHostPorts(zipkinHttpPort);
+    String zipkinEndpoint =
+        "http://host.testcontainers.internal:" + zipkinHttpPort + "/api/v2/spans";
+
+    // TODO: read prior JAVA_OPTS from base layer
```

Comment:
a testcontainers whizz will help on this :)

Reply:
not really possible unless you start a container for the image and do `echo $JAVA_OPTS`. Consider adding `$EXTRA_JAVA_OPTS` to the base layer image or similar

Reply:
its weird.. when browsing I thought I saw someone inspecting a dockerfile to scrape it for something like this... ex the equiv of docker inspect... config.env.JAVA_OPTS

Reply:
well, you can try `DockerClientFactory.instance().client().inspectImageCmd()`

But it obviously won't work if the startup script sets them at runtime, so I would recommend not to rely on inspect as it is not bullet proof

Reply:
ahh right it was here added a commit https://github.com/MicroShed/microshed-testing/blob/6d5b990aec77a47938dba61877840dbf19507343/modules/testcontainers/src/main/java/org/microshed/testing/testcontainers/ApplicationContainer.java#L664

Reply:
yeah this is an image we control and all of them define this in the Dockerfile. good point though.